File: frontend/src/app/shared/validators/cnpj.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * CNPJ Validator - Validates Brazilian CNPJ with check digits
 * Story 3.1: Supplier Management
 */
export function cnpjValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const cnpj = control.value;

    if (!cnpj) {
      return null; // Don't validate empty values (use Validators.required for that)
    }

    // Remove formatting
    const cnpjClean = cnpj.replace(/\D/g, '');

    // Check length
    if (cnpjClean.length !== 14) {
      return { cnpj: { message: 'CNPJ deve ter 14 dígitos' } };
    }

    // Check if all digits are the same
    if (/^(\d)\1{13}$/.test(cnpjClean)) {
      return { cnpj: { message: 'CNPJ inválido' } };
    }

    // Validate check digits
    if (!isValidCnpj(cnpjClean)) {
      return { cnpj: { message: 'CNPJ inválido' } };
    }

    return null;
  };
}

function isValidCnpj(cnpj: string): boolean {
  try {
    // First check digit
    const weights1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
    let sum = 0;
    for (let i = 0; i < 12; i++) {
      sum += parseInt(cnpj.charAt(i)) * weights1[i];
    }
    const digit1 = (sum % 11 < 2) ? 0 : 11 - (sum % 11);

    if (parseInt(cnpj.charAt(12)) !== digit1) {
      return false;
    }

    // Second check digit
    const weights2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
    sum = 0;
    for (let i = 0; i < 13; i++) {
      sum += parseInt(cnpj.charAt(i)) * weights2[i];
    }
    const digit2 = (sum % 11 < 2) ? 0 : 11 - (sum % 11);

    return parseInt(cnpj.charAt(13)) === digit2;

  } catch (e) {
    return false;
  }
}

/**
 * Format CNPJ to standard format: 00.000.000/0000-00
 */
export function formatCnpj(cnpj: string): string {
  if (!cnpj) {
    return '';
  }

  const cnpjClean = cnpj.replace(/\D/g, '');

  if (cnpjClean.length !== 14) {
    return cnpj;
  }

  return `${cnpjClean.substring(0, 2)}.${cnpjClean.substring(2, 5)}.${cnpjClean.substring(5, 8)}/${cnpjClean.substring(8, 12)}-${cnpjClean.substring(12)}`;
}

/**
 * Clean CNPJ formatting
 */
export function cleanCnpj(cnpj: string): string {
  return cnpj ? cnpj.replace(/\D/g, '') : '';
}

File: backend/src/main/java/com/estoquecentral/shared/validator/CnpjValidator.java
package com.estoquecentral.shared.validator;

/**
 * CNPJ (Cadastro Nacional da Pessoa Jurídica) validator for Brazilian company tax IDs.
 *
 * Validates CNPJ format and check digits according to Brazilian government rules.
 *
 * Story 4.1: Customer Management - AC4 (CNPJ validation)
 */
public class CnpjValidator {

    private static final int CNPJ_LENGTH = 14;

    /**
     * Validates a CNPJ string.
     *
     * @param cnpj the CNPJ to validate (with or without formatting)
     * @return true if CNPJ is valid, false otherwise
     */
    public static boolean isValid(String cnpj) {
        if (cnpj == null || cnpj.isEmpty()) {
            return false;
        }

        // Remove formatting characters
        cnpj = cnpj.replaceAll("[^0-9]", "");

        // Check length
        if (cnpj.length() != CNPJ_LENGTH) {
            return false;
        }

        // Check for known invalid CNPJs (all digits the same)
        if (cnpj.matches("(\\d)\\1{13}")) {
            return false;
        }

        // Validate first check digit
        int[] multiplicadores1 = {5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};
        int soma = 0;
        for (int i = 0; i < 12; i++) {
            soma += Character.getNumericValue(cnpj.charAt(i)) * multiplicadores1[i];
        }
        int digito1 = (soma % 11 < 2) ? 0 : 11 - (soma % 11);

        if (Character.getNumericValue(cnpj.charAt(12)) != digito1) {
            return false;
        }

        // Validate second check digit
        int[] multiplicadores2 = {6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};
        soma = 0;
        for (int i = 0; i < 13; i++) {
            soma += Character.getNumericValue(cnpj.charAt(i)) * multiplicadores2[i];
        }
        int digito2 = (soma % 11 < 2) ? 0 : 11 - (soma % 11);

        return Character.getNumericValue(cnpj.charAt(13)) == digito2;
    }

    /**
     * Formats a CNPJ string to the standard format: 00.000.000/0000-00
     *
     * @param cnpj the CNPJ to format (digits only)
     * @return formatted CNPJ string
     */
    public static String format(String cnpj) {
        if (cnpj == null || cnpj.isEmpty()) {
            return cnpj;
        }

        cnpj = cnpj.replaceAll("[^0-9]", "");

        if (cnpj.length() != CNPJ_LENGTH) {
            return cnpj;
        }

        return String.format("%s.%s.%s/%s-%s",
            cnpj.substring(0, 2),
            cnpj.substring(2, 5),
            cnpj.substring(5, 8),
            cnpj.substring(8, 12),
            cnpj.substring(12, 14)
        );
    }

    /**
     * Removes formatting from a CNPJ string (returns digits only).
     *
     * @param cnpj the CNPJ to clean
     * @return CNPJ with digits only
     */
    public static String cleanFormat(String cnpj) {
        if (cnpj == null) {
            return null;
        }
        return cnpj.replaceAll("[^0-9]", "");
    }
}
